### PR TITLE
Fix/enabling validator with empty value

### LINKED ISSUE #283: Impossibility to create new product

### PR DESCRIPTION
**Pull request by maximeroucher**
Close #283

### CODE REVIEW

**Review by guillaumetavernier — APPROVED** [review]

File: lib/tools/ui/widgets/text_entry.dart [hunk 1]
```diff
@@ -78,6 +78,10 @@ class TextEntry extends StatelessWidget {
       ),
       validator: (value) {
```
Comment (julien4215) [inline lib/tools/ui/widgets/text_entry.dart:79]: Even if it works, I think the validator could be cleaner. I would prefer something like this : 

```suggestion
      validator: (value) {
        if (value == null || value.isEmpty) {
          if (canBeEmpty) {
            return null;
          } else {
            return noValueError;
          }
        }

        if (isInt) {
          final intValue = int.tryParse(value);
          if (intValue == null || intValue < 0) {
            return TextConstants.invalidNumber;
          }
        }

        if (isDouble) {
          final doubleValue = double.tryParse(value.replaceAll(',', '.'));
          if (doubleValue == null || doubleValue < 0) {
            return TextConstants.invalidNumber;
          }
        }

        validator?.call(value);
        return null;
      },
```

**Review by julien4215 — CHANGES_REQUESTED** [review]

File: lib/tools/ui/widgets/text_entry.dart [hunk 2]
```diff
@@ -77,26 +77,33 @@ class TextEntry extends StatelessWidget {
         ),
       ),
       validator: (value) {
-        if (canBeEmpty) {
-          if (value == null || value.isEmpty) {
+        if (value == null || value.isEmpty) {
+          if (canBeEmpty) {
             return null;
+          } else {
+            return noValueError;
```
Comment (maximeroucher) [inline lib/tools/ui/widgets/text_entry.dart:84]: I think that you are recreating the issue I intend to solve by passing `validator?.call(value);` here

File: lib/tools/ui/widgets/text_entry.dart [hunk 3]
```diff
@@ -77,26 +77,33 @@ class TextEntry extends StatelessWidget {
...
+          }
         }
-        final doubleValue = double.tryParse(value.replaceAll(',', '.'));
-        if (isDouble && (doubleValue == null || doubleValue < 0)) {
-          return TextConstants.invalidNumber;
+
+        if (validator == null) {
+          return null;
+        } else {
+          return validator!(value);
```
Comment (maximeroucher) [inline lib/tools/ui/widgets/text_entry.dart:105]: Go for `validator?.call(value);`, it does the same thing as those four lines

File: lib/tools/ui/widgets/text_entry.dart (hunk 2 above)
Comment (julien4215) [reply lib/tools/ui/widgets/text_entry.dart:84]: `validator?.call(value)` is called at the end of the validator function.

File: lib/tools/ui/widgets/text_entry.dart (hunk 3 above)
Comment (julien4215) [reply lib/tools/ui/widgets/text_entry.dart:105]: It doesn't. You need these two lines to achieve the same thing.
```
validator?.call(value);
return null;
```
I just thought that would make it easier to understand with the 4 lines.

**Review by guillaumetavernier — APPROVED** [review]

**Review by julien4215 — APPROVED** [review]

**Pull request merged**